p2p/batchwriter.go

```diff
@@ -0,0 +1,372 @@
+package p2p
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/iotexproject/iotex-proto/golang/iotexrpc"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/iotexproject/iotex-core/pkg/lifecycle"
+	"github.com/iotexproject/iotex-core/pkg/log"
+)
+
+const (
+	_bufferLength    = 500
+	_maxWriters      = 5000
+	_cleanupInterval = 15 * time.Minute
+)
+
+type batchOption func(cfg *writerConfig)
+
+func withInterval(t time.Duration) batchOption {
+	return func(cfg *writerConfig) {
+		cfg.msgInterval = t
+	}
+}
+
+func withSizeLimit(limit uint64) batchOption {
+	return func(cfg *writerConfig) {
+		cfg.sizeLimit = limit
+	}
+}
+
+type (
+	batchManager struct {
+		mu                  sync.RWMutex
+		writerMap           map[batchID]*batchWriter
+		outputQueue         chan *batchMessage // assembled message queue for external reader
+		assembleQueue       chan *batch        // batch queue which collects batches sent from writers
+		batchMessageHandler messageOutbound
+		cancelHanlder       context.CancelFunc
+	}
+
+	batchID uint64
+
+	messageOutbound func(*batchMessage) error
+)
+
+func newBatchManager(handler messageOutbound) *batchManager {
```

Comment:
The `batchManager` component couldn't be moved outside of `p2p` module because message serialization is done in the `p2p.convertAppMsg()`.